Alerting: Add rule_matcher filter to Prometheus rules API

alexander-akhmetov opened the pull request:
**What is this feature?**

Add `rule_matcher` filter to the Prometheus-compatible list rules API: `/api/prometheus/grafana/api/v1/rules`. It allows to filter rules by static labels (not by alert instance labels).

**Special notes for your reviewer:**

  - Equality (`=`) and inequality (`!=`) matchers are pushed down to the database. Regex matchers (`=~`, `!~`) are applied in-memory at the API layer.
  - SQLite: Uses GLOB pattern matching
  - MySQL / PostgreSQL: Use JSON functions to compare label values


Example: https://ephemeral15111821115297alexan.grafana-dev.net/alerting/list?search=label:gen_label_1%3Damazing_dragonfly

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

alexander-akhmetov committed:
Alerting: Add rule_label filter to Prometheus rules API

alexander-akhmetov commented:
/deploy-to-hg

alexander-akhmetov commented:
/deploy-to-hg

alexander-akhmetov commented:
/deploy-to-hg

alexander-akhmetov committed:
Fetch less rules when rule_label filter is used

laurenashleigh commented on public/app/features/alerting/unified/rule-list/hooks/grafanaFilter.ts:
```diff
@@ -85,6 +100,7 @@ export function getGrafanaFilter(filterState: Partial<RulesFilter>) {
     dashboardUid: ruleFilterConfig.dashboardUid ? undefined : normalizedFilterState.dashboardUid,
     searchGroupName: groupFilterConfig.groupName ? undefined : normalizedFilterState.groupName,
     datasources: ruleFilterConfig.dataSourceNames ? undefined : datasourceUids,
+    ruleLabels,
```
Does this `ruleLabels` also need to be added to the FE GrafanaPromRulesOptions? (in public/app/features/alerting/unified/api/prometheusApi.ts)

laurenashleigh commented on public/app/features/alerting/unified/rule-list/hooks/grafanaFilter.ts:
```diff
@@ -75,6 +76,20 @@ export function getGrafanaFilter(filterState: Partial<RulesFilter>) {
+        try {
+          return JSON.stringify(parseMatcher(label));
+        } catch {
+          return undefined;
```
Is it worth having a `console.warn` here for debugging?

alexander-akhmetov committed:
Log error when label matcher parsing fails

alexander-akhmetov committed:
Add plugin rule to the test

alexander-akhmetov committed:
Fix linter

konrad147 committed:
Alerting: Add test coverage for rule label backend filtering

Adds test to verify that ruleLabels are correctly included in backend
filter when alertingUIUseBackendFilters feature toggle is enabled.

Also updates hasGrafanaClientSideFilters tests to reflect that labels
are now backend-filtered when the feature toggle is on.

konrad147 reviewed (approved):
LGTM!

alexander-akhmetov committed:
Make empty label matcher to follow Prometheus semantics

alexander-akhmetov committed:
Rename rule_label to rule_matcher

titolins commented on pkg/services/ngalert/api/prometheus/api_prometheus.go:
```diff
@@ -841,41 +877,30 @@ func PrepareRuleGroupStatuses(log log.Logger, store ListAlertRulesStore, opts Ru
 	limitRulesPerGroup := getInt64WithDefault(opts.Query, "limit_rules", -1)
 	limitAlertsPerRule := getInt64WithDefault(opts.Query, "limit_alerts", -1)
-	matchers, err := getMatchersFromQuery(opts.Query)
+	matchers, err := getMatchersFromQuery(opts.Query, "matcher")
```
maybe it's worth defining constants for those params (`matcher`/`rule_(label|matcher)`). wdyt? would help with context as well (instance vs rule matchers).

alexander-akhmetov committed:
Merge branch 'main' of github.com:grafana/grafana into alexander-akhmetov/rule-label-filter

titolins commented on pkg/services/ngalert/api/prometheus/api_prometheus.go:
```diff
@@ -1100,6 +1138,30 @@ func filterRulesByHealth(ruleGroup *apimodels.RuleGroup, withHealthFast map[stri
 }
 
+func filterRulesByLabelMatchers(ruleGroup *apimodels.RuleGroup, matchers labels.Matchers) {
+	if len(matchers) == 0 {
```
probably doesn't change the result, but might make some small difference in the performance if we filtered out the store matchers before no?

titolins reviewed (approved):
Really nice! My only concern here would be making sure the different SQL dialects work correctly, but I'm assuming you're testing them 👍

titolins commented on pkg/services/ngalert/api/prometheus/api_prometheus.go:
```diff
@@ -642,21 +670,30 @@ func PrepareRuleGroupStatusesV2(log log.Logger, store ListAlertRulesStoreV2, opt
 	}
 	span.SetAttributes(attribute.Int("matcher_count", len(matchers)))
 
+	ruleLabelMatchers, err := getMatchersFromQuery(opts.Query, "rule_label")
```
changed already 👍

alexander-akhmetov committed:
review fixes
